smm/lotus_adapter.go

```diff
@@ -0,0 +1,253 @@
+        case changes := <-adapter.headChanges:
+            stateChanges, err := stateChangesFromHeadChanges(changes)
+            if err != nil {
+                // TODO: Log, nothing else to do
+            }
+            for _, stateChange := range stateChanges {
+                adapter.listener.OnChainStateChanged(stateChange)
+            }
+
+        case <-adapter.context.Done():
+            return
+        }
+    }
+}
+
+func (adapter lotusAdapter) callMinerActorMethod(ctx context.Context, method uint64, params typegen.CBORMarshaler) (cid.Cid, error) {
+    payload, aerr := actors.SerializeParams(params)
+    if aerr != nil {
+        return cid.Undef, aerr
+    }
+    msg := types.Message{
+        To:       adapter.actor,
+        From:     adapter.miner,
+        Method:   method,
+        Params:   payload,
+        // TODO: Add ability to control these 'costs'
+        Value:    types.NewInt(0),
+        GasLimit: types.NewInt(1000000),
+        GasPrice: types.NewInt(1),
+    }
+    signedMsg, err := adapter.fullNode.MpoolPushMessage(ctx, &msg)
+    if err != nil {
+        return cid.Undef, err
+    }
+    return signedMsg.Cid(), nil
+}
+
+func (adapter lotusAdapter) Start() {
+    go adapter.eventHandler()
+}
+
+func (adapter lotusAdapter) MostRecentState(ctx context.Context) (StateKey, error) {
+    ts, err := adapter.fullNode.ChainHead(ctx)
+    if err != nil {
+        return "", err
+    }
+    tsk := types.NewTipSetKey(ts.Cids()...)
+    return StateKey(tsk.Bytes()), nil
+}
+
+func (adapter lotusAdapter) GetMinerState(ctx context.Context, stateKey StateKey) (*MinerChainState, error) {
+    state := new(MinerChainState)
+    ts, err := statekey2tipset(ctx, stateKey, adapter.fullNode)
+    if err != nil {
+        return nil, err
+    }
+    sectors, err := adapter.fullNode.StateMinerSectors(ctx, adapter.miner, ts)
+    if err != nil {
+        return nil, err
+    }
+    provingSectors, err := adapter.fullNode.StateMinerProvingSet(ctx, adapter.miner, ts)
+    if err != nil {
+        return nil, err
+    }
+    state.Sectors = make(map[uint64]api.ChainSectorInfo)
+    for _, sectorInfo := range sectors {
+        state.Sectors[sectorInfo.SectorID] = *sectorInfo
+    }
+    for _, sectorInfo := range provingSectors {
+        state.ProvingSet[sectorInfo.SectorID] = struct{}{}
+    }
+    state.Address = Address(adapter.miner.String())
+    state.PreCommittedSectors = make(map[uint64]api.ChainSectorInfo)
+    state.StagedCommittedSectors = make(map[uint64]api.ChainSectorInfo)
+    state.ProvingSet = make(map[uint64]struct{})
+    return state, nil
+}
+
+func (adapter lotusAdapter) GetRandomness(ctx context.Context, stateKey StateKey, offset uint) ([]byte, error) {
+    ts, err := statekey2tipset(ctx, stateKey, adapter.fullNode)
+    if err != nil {
+        return nil, err
+    }
+    tsk := types.NewTipSetKey(ts.Cids()...)
+    return adapter.fullNode.ChainGetRandomness(ctx, tsk, ts.Blocks()[0].Tickets, int(offset))
```

Comment:
I believe the tickets parameter should be nil here - it's interpreted as a ticket chain extending the blockchain (e.g. for null rounds). I'm pretty sure passing tickets from the block is wrong. Which means you don't need to load the tipset at all, and can just use the key.